sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -742,6 +742,23 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelpe
  * equivalent [[Literal]] values.
  */
 object ConstantFolding extends Rule[LogicalPlan] {
+  private def isAssociativelyFoldable(e: Expression): Boolean =
```

Comment:
Oh, that could be. 

There is some difference on level of granulity.

Join-related optimizers might be improved later to cost-based optimizers while ConstantFolder optimizer is just about removing constants on a single expression.

Do you think it is a good idea to put the different levels of concerns together?

I can do this in any way you decide. :)